src/devices/switch.cpp

```diff
@@ -28,28 +28,134 @@ uuid::log::Logger Switch::logger_ {
+    if (export_values(json)) {
+        print_value_json(shell, F("activated"), nullptr, F_(activated), nullptr, json);
+        print_value_json(shell, F("flowTemp"), F_(2spaces), F_(flowTemp), F_(degrees), json);
+        print_value_json(shell, F("status"), nullptr, F_(status), nullptr, json);
+    }
 }
 
 // publish values via MQTT
 void Switch::publish_values(JsonObject & json, bool force) {
+    if (Mqtt::mqtt_format() == Mqtt::Format::HA) {
+        if (!mqtt_ha_config_ || force) {
+            register_mqtt_ha_config();
+            return;
+        }
+    }
+    StaticJsonDocument<EMSESP_MAX_JSON_SIZE_SMALL> doc;
+    JsonObject                                     json_data = doc.to<JsonObject>();
+    if (export_values(json_data)) {
+        Mqtt::publish(F("switch_data"), doc.as<JsonObject>());
+    }
 }
 
 // export values to JSON
 bool Switch::export_values(JsonObject & json) {
+
+    if (Helpers::hasValue(flowTemp_)) {
+        char s[7];
+        json["activated"] = Helpers::render_value(s, activated_, EMS_VALUE_BOOL);
+    }
+
+    if (Helpers::hasValue(flowTemp_)) {
+        json["flowTemp"] = (float)flowTemp_ / 10;
+    }
+
+    if (Helpers::hasValue(flowTemp_)) {
+        json["status"] = status_;
+    }
+
     return true;
 }
 
 // check to see if values have been updated
 bool Switch::updated_values() {
+    if (changed_) {
+        changed_ = false;
+        return true;
+    }
     return false;
 }
 
+// publish config topic for HA MQTT Discovery
+void Switch::register_mqtt_ha_config() {
+    if (!Mqtt::connected()) {
+        return;
+    }
+
+    // if we don't have valid values for this HC don't add it ever again
+    if (!Helpers::hasValue(flowTemp_)) {
+        return;
+    }
+
+    // Create the Master device
+    StaticJsonDocument<EMSESP_MAX_JSON_SIZE_HA_CONFIG> doc;
+
+    char name[10];
+    snprintf_P(name, sizeof(name), PSTR("Switch"));
+    doc["name"] = name;
+
+    char uniq_id[10];
+    snprintf_P(uniq_id, sizeof(uniq_id), PSTR("switch"));
+    doc["uniq_id"] = uniq_id;
+
+    doc["ic"] = F("mdi:home-thermometer-outline");
+
+    char stat_t[50];
+    snprintf_P(stat_t, sizeof(stat_t), PSTR("%s/switch_data"), System::hostname().c_str());
+    doc["stat_t"] = stat_t;
+
+    doc["val_tpl"] = F("{{value_json.type}}"); // HA needs a single value. We take the type which is wwc or hc
+
+    JsonObject dev = doc.createNestedObject("dev");
+    dev["name"]    = F("EMS-ESP Switch");
+    dev["sw"]      = EMSESP_APP_VERSION;
+    dev["mf"]      = this->brand_to_string();
+    dev["mdl"]     = this->name();
+    JsonArray ids  = dev.createNestedArray("ids");
+    ids.add("ems-esp-switch");
+
+    Mqtt::publish_retain(F("homeassistant/sensor/ems-esp/switch/config"), doc.as<JsonObject>(), true); // publish the config payload with retain flag
+    Mqtt::register_mqtt_ha_sensor(PSTR("switch"), nullptr, F_(flowTemp), this->device_type(), "activated", nullptr, nullptr);
```

Comment:
add to locale_en and use F_(...) ?

Reply:
Ups, these register_mqtt_ha_sensor have some more copy&past errors.

Reply:
should be fixed now.